core/local/local.go

```diff
@@ -23,527 +23,363 @@ package local
 	}
-}
 
-func (e *Executor) Run(parent context.Context, engineOut chan<- stats.SampleContainer) (reterr error) {
-	e.runLock.Lock()
-	defer e.runLock.Unlock()
+	vusFmt := pb.GetFixedLengthIntFormat(int64(e.maxPossibleVUs))
+	return fmt.Sprintf(
+		"%s, "+vusFmt+"/"+vusFmt+" VUs, %d complete and %d incomplete iterations",
+		status, e.state.GetCurrentlyActiveVUsCount(), e.state.GetInitializedVUsCount(),
+		e.state.GetFullIterationCount(), e.state.GetPartialIterationCount(),
+	)
+}
 
-	if e.Runner != nil && e.runSetup {
-		if err := e.Runner.Setup(parent, engineOut); err != nil {
-			return err
+// Init concurrently initializes all of the planned VUs and then sequentially
+// initializes all of the configured schedulers.
+func (e *Executor) Init(ctx context.Context, engineOut chan<- stats.SampleContainer) error {
+	logger := e.logger.WithField("phase", "local-executor-init")
+
+	vusToInitialize := lib.GetMaxPlannedVUs(e.executionPlan)
+	logger.WithFields(logrus.Fields{
+		"neededVUs":       vusToInitialize,
+		"schedulersCount": len(e.schedulers),
+	}).Debugf("Start of initialization")
+
+	doneInits := make(chan error, vusToInitialize) // poor man's early-return waitgroup
+	//TODO: make this an option?
+	initConcurrency := runtime.NumCPU()
+	limiter := make(chan struct{}, initConcurrency)
+	subctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	initPlannedVU := func() {
+		newVU, err := e.initVU(ctx, logger, engineOut)
+		if err == nil {
+			e.state.AddInitializedVU(newVU)
+			<-limiter
 		}
+		doneInits <- err
 	}
 
-	ctx, cancel := context.WithCancel(parent)
-	vuFlow := make(chan int64)
-	e.lock.Lock()
-	vuOut := e.vuOut
-	iterDone := e.iterDone
-	e.ctx = ctx
-	e.flow = vuFlow
-	e.lock.Unlock()
-
-	var cutoff time.Time
-	defer func() {
-		if e.Runner != nil && e.runTeardown {
-			err := e.Runner.Teardown(parent, engineOut)
-			if reterr == nil {
-				reterr = err
-			} else if err != nil {
-				reterr = fmt.Errorf("teardown error %#v\nPrevious error: %#v", err, reterr)
-			}
-		}
-
-		close(vuFlow)
-		cancel()
-
-		e.lock.Lock()
-		e.ctx = nil
-		e.vuOut = nil
-		e.flow = nil
-		e.lock.Unlock()
-
-		wait := make(chan interface{})
-		go func() {
-			e.wg.Wait()
-			close(wait)
-		}()
-
-		for {
+	go func() {
+		for vuNum := uint64(0); vuNum < vusToInitialize; vuNum++ {
 			select {
-			case <-iterDone:
-				// Spool through all remaining iterations, do not emit stats since the Run() is over
-			case newSampleContainer := <-vuOut:
-				if cutoff.IsZero() {
-					engineOut <- newSampleContainer
-				} else if csc, ok := newSampleContainer.(stats.ConnectedSampleContainer); ok && csc.GetTime().Before(cutoff) {
-					engineOut <- newSampleContainer
-				} else {
-					for _, s := range newSampleContainer.GetSamples() {
-						if s.Time.Before(cutoff) {
-							engineOut <- s
-						}
-					}
-				}
-			case <-wait:
-			}
-			select {
-			case <-wait:
-				close(vuOut)
+			case limiter <- struct{}{}:
+				go initPlannedVU()
```

Comment:
It would be better (probably not by a lot) to run initConcurrency goroutines which read from range over `limiter`  and just execute the `initPlannedVu`. you will also need to close limiter. 
This is better primarily because it reuses goroutines. In the particular case of initializing VUs ... on my machine 1000 VUs with practically no code take something like 6G to initialize so I doubt the improvement will be significant ... actually  I measured it and I couldn't ... reliably detect improvement. 

Below is my diff 
```
--- a/core/local/local.go
+++ b/core/local/local.go
@@ -186,24 +186,26 @@ func (e *Executor) Init(ctx context.Context, engineOut chan<- stats.SampleContai
        doneInits := make(chan error, vusToInitialize) // poor man's early-return waitgroup
        //TODO: make this an option?
        initConcurrency := runtime.NumCPU()
-       limiter := make(chan struct{}, initConcurrency)
+       limiter := make(chan struct{})
        subctx, cancel := context.WithCancel(ctx)
        defer cancel()
-
-       initPlannedVU := func() {
-               newVU, err := e.initVU(ctx, logger, engineOut)
-               if err == nil {
-                       e.state.AddInitializedVU(newVU)
-                       <-limiter
-               }
-               doneInits <- err
+       for i := 0; i < initConcurrency; i++ {
+               go func() {
+                       for range limiter {
+                               newVU, err := e.initVU(ctx, logger, engineOut)
+                               if err == nil {
+                                       e.state.AddInitializedVU(newVU)
+                               }
+                               doneInits <- err
+                       }
+               }()
        }

        go func() {
+               defer close(limiter)
                for vuNum := uint64(0); vuNum < vusToInitialize; vuNum++ {
                        select {
                        case limiter <- struct{}{}:
-                               go initPlannedVU()
                        case <-subctx.Done():
                                return
                        }

```

Reply:
So... should I change it or not? :grinning:

Reply:
yes, I think it is worth the change ... in cases where we have high CPU count with high VUs this would be worth it

Reply:
Resolved by https://github.com/loadimpact/k6/pull/1007/commits/ab2fab6ded46ddfbffd9eebce6580d3b75307f2c